File: linked_list/remove_nth_node_from_end_of_list.19.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

 // To remove a node, we need access to the node BEFORE it.
// if n = length, we need to delete the head node, and there is no node before it in the linked list.
// so we need to create a dummy node before the head to handle this case.
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        // traverse and get the length
        ListNode* current = head;
        int length = 0;
        while (current != nullptr) {
            ++length;
            current = current->next;
        }
        
        ListNode* dummy = new ListNode(0);
        dummy->next = head;
        current = dummy;
        for (int i = 0; i < length - n; ++i) {
            current = current->next;
        } // reach the node before the node to be deleted
        ListNode* node_to_delete = current->next; // prevent memory leak?
        current->next = current->next->next;
        delete node_to_delete;

        ListNode* new_list = dummy->next;
        delete dummy;
        return new_list;
    }
};

// e.g. length = 5, n = 2
// --> delete the 4th node, whose index is length - n = 3.
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        // Traverse and get the length
        ListNode* current = head;
        int length = 0;
        while (current != nullptr) {
            ++length;
            current = current->next;
        }
        
        // edge case: remove the head
        if (length == n) {
            ListNode* new_head = head->next;
            delete head;
            return new_head;
        }
        // general case:
        current = head;
        for (int i = 0; i < length - n - 1; i++) {
            current = current->next;
        }
        current->next = current->next->next;
        return head;
    }
};

File: linked_list/remove_nth_node_from_end_of_list.19.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

# Dummy node 解决的不是"链表"问题，而是"前驱节点不存在"的问题。
# O(l) time
# O(1) space
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        # length
        l = 0
        curr = head
        while curr:
            l += 1
            curr = curr.next
        dummy = ListNode(-1)
        dummy.next = head
        curr = dummy
        i = 0
        while i < l-n:
            curr = curr.next
            i += 1
        # IMPORTANT: need to stop at the node before the node to be deleted
        curr.next = curr.next.next
        return dummy.next

# Two pointers
def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
    dummy = ListNode(-1)
    dummy.next = head
    fast = slow = dummy

    # fast 先走 n+1 步
    for _ in range(n + 1):
        fast = fast.next

    # 同步前进直到 fast 到尾
    while fast:
        fast = fast.next
        slow = slow.next

    # slow 此时在待删节点的前驱
    slow.next = slow.next.next
    return dummy.next